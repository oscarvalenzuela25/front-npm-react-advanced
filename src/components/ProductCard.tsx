import styles from '../styles/styles.module.css';
import useProduct from '../hooks/useProduct';
import React, { createContext, CSSProperties } from 'react';
import {
  ProductContextProps,
  Product,
  OnChangesArgs,
  InitialValues,
  ProductCardHandlers,
} from '../interfaces/interfaces';

export interface Props {
  // Patrones de componentes
  // children: ReactElement | ReactElement[];
  product: Product;
  className?: string;
  style?: CSSProperties;
  onChange?: ({ product, count }: OnChangesArgs) => void;
  value?: number;

  //State initializer
  children: (args: ProductCardHandlers) => JSX.Element;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};

export default ProductCard;
