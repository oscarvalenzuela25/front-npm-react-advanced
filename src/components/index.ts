export { default as ProductCard } from '../components/ProductCard';
export { ProductTitle } from '../components/ProductTitle';
export { ProductImage } from '../components/ProductImage';
export { ProductButtons } from '../components/ProductButtons';
