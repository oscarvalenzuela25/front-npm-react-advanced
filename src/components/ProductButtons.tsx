import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
  const { counter, increaseBy, maxCount } = useContext(ProductContext);

  // El useCallback en si es para tener una funcion que se ejecuta con dependencias.
  // Imagina que en vez del callback es un useEffect con las dependencias y estas actualizando un estado
  // Aqui actualiza la funcion basado en las dependencias
  const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        {' '}
        -{' '}
      </button>
      <div className={styles.countLabel}> {counter} </div>
      <button
        className={
          isMaxReached()
            ? `${styles.disabled} ${styles.buttonAdd}`
            : styles.buttonAdd
        }
        onClick={() => !isMaxReached() && increaseBy(1)}
      >
        {' '}
        +{' '}
      </button>
    </div>
  );
};
