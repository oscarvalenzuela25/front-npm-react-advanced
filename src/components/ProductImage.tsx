import React, { CSSProperties, useContext } from 'react';
import noImage from '../assets/no-image.jpg';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img, className, style }: Props) => {
  const { product } = useContext(ProductContext);
  let imgToShow: string;
  if (img) {
    imgToShow = img;
  } else {
    imgToShow = product.img || noImage;
  }

  return (
    <img
      className={`${styles.productImg} ${className}`}
      style={style}
      src={imgToShow}
      alt="Coffee Mug"
    />
  );
};
