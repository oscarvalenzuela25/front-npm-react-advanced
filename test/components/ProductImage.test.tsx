import React from 'react';
import rendered from 'react-test-renderer';
import ProductCard from '../../src/components/ProductCard';
import { product2 } from './data/products';
import { ProductImage } from './../../src/components/ProductImage';

describe('ProductImage', () => {
  test('Debe mostrar el componente con una img', () => {
    const wrapper = rendered.create(<ProductImage img={product2.img} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe mostrar el componente con la img del producto', () => {
    const wrapper = rendered.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
