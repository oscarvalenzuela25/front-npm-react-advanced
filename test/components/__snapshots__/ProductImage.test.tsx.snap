// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostrar el componente con la img del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffee Mug"
    className="productImg undefined"
    src="./coffee-mug2.png"
  />
</div>
`;

exports[`ProductImage Debe mostrar el componente con una img 1`] = `
<img
  alt="Coffee Mug"
  className="productImg undefined"
  src="./coffee-mug2.png"
/>
`;
